README.md
```bash      
 ███████╗██╗  ██╗ █████╗ ██████╗  ██████╗ ██╗    ██╗      ██████╗  ██████╗ ██╗  ██╗
 ██╔════╝██║  ██║██╔══██╗██╔══██╗██╔═══██╗██║    ██║      ██╔══██╗██╔═══██╗╚██╗██╔╝
 ███████╗███████║███████║██║  ██║██║   ██║██║ █╗ ██║█████╗██████╔╝██║   ██║ ╚███╔╝ 
 ╚════██║██╔══██║██╔══██║██║  ██║██║   ██║██║███╗██║╚════╝██╔══██╗██║   ██║ ██╔██╗ 
 ███████║██║  ██║██║  ██║██████╔╝╚██████╔╝╚███╔███╔╝      ██████╔╝╚██████╔╝██╔╝ ██╗
 ╚══════╝╚═╝  ╚═╝╚═╝  ╚═╝╚═════╝  ╚═════╝  ╚══╝╚══╝       ╚═════╝  ╚═════╝ ╚═╝  ╚═╝
      
                Lightweight Hypervisor-Based Kernel Protector
```

# 1. Notice
# 1.1. Presentations and Papers
Shadow-box is a lightweight and practical kernel protector, and it was introduced at security conferences below.
 - [Black Hat Asia 2017](https://www.blackhat.com/asia-17/briefings.html#myth-and-truth-about-hypervisor-based-kernel-protector-the-reason-why-you-need-shadow-box) and [Black Hat Asia Arsenal 2017](https://www.blackhat.com/asia-17/arsenal.html#shadow-box-lightweight-hypervisor-based-kernel-protector)
 - [HITBSecConf 2017](http://conference.hitb.org/hitbsecconf2017ams/sessions/shadowbox-the-practical-and-omnipotent-sandbox/)

You can watch the demo videos below.
 - [Demo 1](https://youtu.be/3_cFDVHWCXA): It shows that if you use only kernel-level protection mechanism such as page write-protect, rootkits can neutralize it. 
 - [Demo 2](https://youtu.be/s7iZYg4vP4E): It shows that if you use kernel-level protection mechanism with Shadow-box, then rootkits cannot neutralize it and cannot work. 

# 1.2. Contributions
We always welcome your contributions. Issue report, bug fix, new feature implementation, anything is alright. Feel free to send us. 

# 1.3. License
Shadow-box has dual license, GPL v2 and MIT. You can choose any license you want.

# 2. Introduction of Shadow-Box
Shadow-box is a security monitoring framework for operating systems using state-of-the-art virtualization technologies. 
Shadow-box has a novel architecture inspired by a shadow play. We made Shadow-box from scratch, and it is primarily composed of a lightweight hypervisor and a security monitor. 

The lightweight hypervisor, Light-box, efficiently isolates an OS inside a guest machine, and projects static and dynamic kernel objects of the guest into the host machine so that our security monitor in the host can investigate the projected images. 
The security monitor, Shadow-Watcher, places event monitors on static kernel elements and tests security of dynamic kernel elements. 

Shadow-box manipulates address translations from the guest physical address to the host physical address in order to exclude unauthorized accesses to the host and the hypervisor spaces. In that way, Shadow-box can properly introspect the guest operating system and mediate all accesses, even when the operating system is compromised.

# 2.1. Architecture of Shadow-Box
We explain how we designed the Light-box and the Shadow-watcher. It is designed to support a lightweight and practical security monitoring framework using virtualization technologies.

<center> <img src="document/images/architecture.png" width="500"> </center>

We developed a security monitoring framework, Shadow-box that keeps an OS safe by filtering out unauthorized accesses to important kernel elements and defending integrity of kernel elements periodically. Shadow-box relies upon its two subparts: a lightweight hypervisor and a security monitor. The lightweight hypervisor, Light-box, efficiently isolates an OS inside a guest machine, and projects static and dynamic kernel objects of the guest into the host machine, so that our security monitor in the host can investigate the projected images. The security monitor, Shadow-watcher, places event monitors on static kernel elements and tests security of dynamic kernel elements. Running inside the host, it can test the security of the guest without malicious interference even when the guest OS is compromised.

If you want to know more about Shadow-box, please see my presentation and paper at [Black Hat Asia 2017](https://www.blackhat.com/asia-17/briefings.html#myth-and-truth-about-hypervisor-based-kernel-protector-the-reason-why-you-need-shadow-box) and [HITBSecConf 2017](http://conference.hitb.org/hitbsecconf2017ams/sessions/shadowbox-the-practical-and-omnipotent-sandbox/).


# 3. How to Build 
# 3.1. Prepare Kernel Build Environment (Ubuntu 16.04)
Because the Shadow-box protects the code area of kernel, it conflicts with the runtime kernel patch feature (CONFIG_JUMP_LABEL). Therefore, if your kernel uses the runtime kernel patch feature, you should remove the feature. To remove it, you need to set a kernel build environment, change the kernel options, and install. The process is as follows.

```bash
# Prepare kernel source and build environment.
$> apt-get source linux
$> sudo apt-get build-dep linux
$> sudo apt-get install linux-headers-<your_kernel_version> ncurses-dev

# Make new .config file.
$> cd linux-<your_kernel_version>
$> cp /boot/config-<your_kernel_version> .config
$> make menuconfig
# Load the .config file using the "Load" menu and save it to .config using the "Save" menu.

$> sed -i 's/CONFIG_JUMP_LABEL=y/# CONFIG_JUMP_LABEL is not set/g' .config

# Build kernel and modules.
$> make -j8; make modules

# Install kernel and modules.
$> sudo make modules_install
$> sudo make install
``` 

# 3.2. Prepare Shadow-Box and Kernel Symbols
Shadow-box should locate the data structures and functions for kernel integrity verification. These symbols can be found using kallsyms, but all symbols are not exposed to kallsyms. Therefore, Shadow-box uses the System.map file to embed symbols and kernel versions.
How to add symbols and kernel versions to Shadow-box is as follows:

```bash
# Prepare Shadow-box source.
$> git clone https://github.com/kkamagui/shadow-box-for-x86.git
$> cd shadow-box-for-x86

# Prepare kernel symbols.
$> uname -v
#37-Ubuntu SMP Mon Apr 18 18:33:37 UTC 2016			<== Kernel version

# Copy system.map file to kernel_version name.
$> cp /boot/System.map-<your_kernel_version> system.map/"#37-Ubuntu SMP Mon Apr 18 18:33:37 UTC 2016.map"
```

# 3.3. Build Shadow-Box
When the kernel symbol is ready, type "make" command to build the Shadow-box. Then you can find shadow_box.ko in the same directory.
```bash
$> make
$> ls
shadow_box.ko shadow_box.h ...

```

# 4. How to Use
Shadow-box is loadable kernel module (LKM). So, when you need protection, you can load the shadow-box.ko module into the kernel with the insmod command.
```bash
$> sudo insmod shadow-box.ko
```
<center> <img src="document/images/screenshot1.png" width="700"> </center>

After Shadow-box is loaded, you can test the protection mechanisms of Shadow-box using rootkits such as [Adore-ng](https://github.com/kkamagui/adore-ng). If you want to watch demo videos, please follow these links, [Demo 1](https://youtu.be/3_cFDVHWCXA) and [Demo 2](https://youtu.be/s7iZYg4vP4E).

Here is the rootkit detection example. If Shadow-box detects rootkits, Shadow-box writes kernel log messages about rootkit attacks.

```bash
# Load Adore-ng rootkit by command-line.
$> insmod adore_ng.ko
Segmentation fault (codre dumped)

# Show kernel log messages.
$> dmesg
... omitted ...
[    4.182249] Bluetooth: BNEP filters: protocol multicast
[    4.182252] Bluetooth: BNEP socket layer initialized
[   43.615020] shadow_box: module verification failed: signature and/or required key missing - tainting kernel
[   43.625298]      
[   43.625299] ███████╗██╗  ██╗ █████╗ ██████╗  ██████╗ ██╗    ██╗      ██████╗  ██████╗ ██╗  ██╗
[   43.625300] ██╔════╝██║  ██║██╔══██╗██╔══██╗██╔═══██╗██║    ██║      ██╔══██╗██╔═══██╗╚██╗██╔╝
[   43.625301] ███████╗███████║███████║██║  ██║██║   ██║██║ █╗ ██║█████╗██████╔╝██║   ██║ ╚███╔╝ 
[   43.625302] ╚════██║██╔══██║██╔══██║██║  ██║██║   ██║██║███╗██║╚════╝██╔══██╗██║   ██║ ██╔██╗ 
[   43.625303] ███████║██║  ██║██║  ██║██████╔╝╚██████╔╝╚███╔███╔╝      ██████╔╝╚██████╔╝██╔╝ ██╗
[   43.625303] ╚══════╝╚═╝  ╚═╝╚═╝  ╚═╝╚═════╝  ╚═════╝  ╚══╝╚══╝       ╚═════╝  ╚═════╝ ╚═╝  ╚═╝
[   43.625304]      
[   43.625304]                Lightweight Hypervisor-Based Kernel Protector
[   43.625305]      
[   43.633127] Shadow-box: CPU Count 4
[   43.633128] Shadow-box: Booting CPU ID 0
[   43.648616] Shadow-box: Protect Kernel Code Area
[   43.670134] Shadow-box:     [*] Complete
[   43.670135] Shadow-box: Protect Module Code Area
[   43.677168] Shadow-box:     [*] Complete
[   43.677338] Shadow-box: Framework Preinitialize
[   43.678077] Shadow-box:     [*] Complete
[   43.702142] Shadow-box: Framework Initailize
[   43.702180] Shadow-box:     [*] Task count 225
[   43.702181] Shadow-box:     [*] Module count 86
[   43.702182] Shadow-box:     [*] Complete
[   44.793463] Shadow-box: Lock IOMMU
[   44.794091] Shadow-box:     [*] Lock IOMMU complete
[   44.921020] Shadow-box: Execution Complete
[   44.921023] Shadow-box: ErrorCode: 0

#==================================================
# Adore-ng rootkit module is detected.
#==================================================
[   67.081071] Shadow-box: VM [0] Kernel module is loaded. Current PID: 1805, PPID: 1804, process name: insmod, module: adore_ng

[   67.081411]  ▄▄▄      ▓█████▄  ▒█████   ██▀███  ▓█████        ███▄    █   ▄████ 
[   67.081412]  ▒████▄    ▒██▀ ██▌▒██▒  ██▒▓██ ▒ ██▒▓█   ▀        ██ ▀█   █  ██▒ ▀█▒ 
[   67.081413]  ▒██  ▀█▄  ░██   █▌▒██░  ██▒▓██ ░▄█ ▒▒███    ███  ▓██  ▀█ ██▒▒██░▄▄▄░ 
[   67.081414]  ░██▄▄▄▄██ ░▓█▄   ▌▒██   ██░▒██▀▀█▄  ▒▓█  ▄  ▒▒▒  ▓██▒  ▐▌██▒░▓█  ██▓ 
[   67.081415]   ▓█   ▓██▒░▒████▓ ░ ████▓▒░░██▓ ▒██▒░▒████▒      ▒██░   ▓██░░▒▓███▀▒ 
[   67.081415]    ▒▒   ▓▒█░ ▒▒▓  ▒ ░ ▒░▒░▒░ ░ ▒▓ ░▒▓░░░ ▒░ ░      ░ ▒░   ▒ ▒  ░▒   ▒ 
[   67.081416]      ▒   ▒▒ ░ ░ ▒  ▒   ░ ▒ ▒░   ░▒ ░ ▒░ ░ ░  ░      ░ ░░   ░ ▒░  ░   ░ 
[   67.081417] 	   ░   ▒    ░ ░  ░ ░ ░ ░ ▒    ░░   ░    ░            ░   ░ ░ ░ ░   ░ 
[   67.081418] 	         ░  ░   ░        ░ ░     ░        ░  ░               ░       ░ 
[   67.081418] 			            ░                                                       
[   67.081419] Hide process PID 1738

#==================================================
# A memory attack at FFFFFFFF81A2D2400 is detected.
#==================================================
[   67.081448] Shadow-box: VM [0] Memory attack is detected
[   67.081450] Shadow-box: VM [0] Guest Linear: -2120035776, FFFFFFFF81A2D240
[   67.081451] Shadow-box: VM [0] Guest Physical: 44225088, 0000000002A2D240
[   67.081452] Shadow-box: VM [0] virt_to_phys: virt FFFFFFFF81A2D240 phys 0000000002A2D240
[   67.081453] Shadow-box: ErrorCode: 4
[   67.081463] general protection fault: 0000 [#1] SMP 
[   67.081485] Modules linked in: bnep snd_hda_codec_hdmi dell_led snd_hda_codec_realtek snd_hda_codec_generic nls_iso8859_1 snd_hda_intel snd_hda_codec intel_rapl x86_pkg_temp_thermal intel_powerclamp coretemp snd_hda_core kvm_intel snd_hwdep kvm dell_wmi sparse_keymap input_leds joydev snd_pcm irqbypass crct10dif_pclmul crc32_pclmul dcdbas snd_seq_midi snd_seq_midi_event aesni_intel aes_x86_64 lrw gf128mul glue_helper ablk_helper snd_rawmidi cryptd snd_seq serio_raw

#==================================================
# A hidden process, backdoor, is detected.
#==================================================
[   67.081632] Shadow-box: VM [1] Task count is different: expect 220, real 219
[   67.081653]  snd_seq_device
[   67.081659] Shadow-box: VM [1] Task PID: 1738, TGID: 1738, fork name: backdoor, process name: backdoor is hidden

[   67.081692]  snd_timer snd soundcore mei_me shpchp mei hci_uart btbcm btqca 8250_fintek btintel bluetooth intel_lpss_acpi intel_lpss acpi_als kfifo_buf acpi_pad industrialio mac_hid parport_pc ppdev lp parport autofs4 hid_generic usbhid i915_bpo intel_ips i2c_algo_bit drm_kms_helper syscopyarea sysfillrect e1000e psmouse sysimgblt fb_sys_fops ptp drm ahci pps_core libahci wmi video pinctrl_sunrisepoint i2c_hid pinctrl_intel hid fjes
[   67.081843] CPU: 0 PID: 1805 Comm: insmod Tainted: G           OE   4.4.0-21-generic #37-Ubuntu
[   67.081867] Hardware name: Dell Inc. OptiPlex 7040/096JG8, BIOS 1.2.8 01/26/2016
[   67.081887] task: ffff8801535dd880 ti: ffff880153520000 task.ti: ffff880153520000
[   67.081908] RIP: 0010:[<ffffffffc00081ac>]  [<ffffffffc00081ac>] init_module+0x1ac/0x1000 [adore_ng]
[   67.081939] RSP: 0018:ffff880153523c60  EFLAGS: 00010286
[   67.081974] RAX: ffffffff81a2d240 RBX: ffffffff81e11080 RCX: 0000000000050bc5
[   67.081993] RDX: ffffffff8127a080 RSI: ffffffffc0620480 RDI: ffff880155001300
[   67.082012] RBP: ffff880153523c88 R08: 000000000001a080 R09: ffffffff8121be34
[   67.082032] R10: ffffea00019f9a00 R11: 0000000000000000 R12: ffff880150f87300
[   67.082051] R13: 0000000000000000 R14: ffffffffc0008000 R15: ffff8801516802a0
[   67.082071] FS:  00007fabee1ad700(0000) GS:ffff880159c00000(0000) knlGS:0000000000000000
[   67.082093] CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080040033
[   67.082109] CR2: 000055b34445f5b8 CR3: 0000000135b46000 CR4: 00000000003426f0
[   67.082131] DR0: 0000000000000041 DR1: 0000000000000041 DR2: 0000000000000041
[   67.082151] DR3: 0000000000000041 DR6: 0000000000000041 DR7: 0000000000000041
[   67.082157] Shadow-box: ErrorCode: 4
[   67.082180] Stack:
[   67.082186]  ffffc900012b3fff 00000000b870b587 ffffffff81e11080 ffff880150fa4340
[   67.082211]  0000000000000000 ffff880153523d08 ffffffff81002123 ffffea000278af80
[   67.082234]  ffff880153523ce0 0000000000000286 0000000000000001 0000000000000007
[   67.082258] Call Trace:
[   67.082270]  [<ffffffff81002123>] do_one_initcall+0xb3/0x200
[   67.082287]  [<ffffffff811eaeb3>] ? kmem_cache_alloc_trace+0x183/0x1f0
[   67.082308]  [<ffffffff8118c163>] do_init_module+0x5f/0x1cf
[   67.082324]  [<ffffffff81109df7>] load_module+0x1667/0x1c00
[   67.082341]  [<ffffffff811063a0>] ? __symbol_put+0x60/0x60
[   67.082358]  [<ffffffff812126b0>] ? kernel_read+0x50/0x80
[   67.082374]  [<ffffffff8110a5d4>] SYSC_finit_module+0xb4/0xe0
[   67.082391]  [<ffffffff8110a61e>] SyS_finit_module+0xe/0x10
[   67.082408]  [<ffffffff818244f2>] entry_SYSCALL_64_fastpath+0x16/0x71
[   67.082426] Code: ff fe ff 0f 22 c0 49 8b 44 24 18 48 89 15 25 83 61 00 48 c7 c6 80 04 62 c0 48 8b 40 30 48 8b 40 20 48 8b 10 48 89 15 cc 82 61 00 <48> c7 00 f0 9e 61 c0 48 c7 c2 70 90 61 c0 48 8b 3d 67 3e 61 00 
[   67.082539] RIP  [<ffffffffc00081ac>] init_module+0x1ac/0x1000 [adore_ng]
[   67.082559]  RSP <ffff880153523c60>
[   67.089935] ---[ end trace f2a6f5de2615c18a ]---

#==================================================
# A hidden module, adore_ng, is detected.
#==================================================
[   67.120456] Shadow-box: VM [2] Module count is different: expect 87, real 86
[   67.120479] Shadow-box: VM [2] Hidden module name: adore_ng, ptr: ffffffffc061c080
[   67.121298] Shadow-box: ErrorCode: 4
```

# = Caution =
Shadow-box protects kernel code, read-only data, system table, privilege register, etc. from rootkits. So, if you want to use Shadow-box, you should disable some features below.
 * Disable CONFIG_JUMP_LABEL
   * Change kernel config (.config)
 * Disable hibernation and suspend
   * Change system power management setting
 * Disable IRQ remapping for using IOMMU protection feature
   * Insert intremap=off at the end of linux /boot/vmlinuz... line in grub.cfg file


debian/changelog
gooroom-os-protector (3.1.0+5.10.136~1+grm3u1) stable; urgency=medium

  [ jongkyung.woo ]
  * [3ea2707e] Released 5.10.136-1+grm3u1 (5.10.0-17)

 -- Gooroom Autobuilder <jenkins@gooroom.kr>  Tue, 17 Jan 2023 16:24:04 +0900

gooroom-os-protector (3.1.0+5.10.120~1+grm3u3) stable; urgency=medium

  [ jongkyung.woo ]
  * [f0cada8a] Released Version 3.1.0+5.10.120_1+grm3u3 (5.10.120-1+grm3u2)

 -- Gooroom Autobuilder <jenkins@gooroom.kr>  Wed, 07 Sep 2022 17:06:14 +0900

gooroom-os-protector (3.1.0+5.10.120~1+grm3u2) stable; urgency=medium

  [ jongkyung.woo ]
  * [0b93bed2] Modified libarc4 in preload_modules

 -- Gooroom Autobuilder <jenkins@gooroom.kr>  Fri, 15 Jul 2022 19:17:37 +0900

gooroom-os-protector (3.1.0+5.10.120~1+grm3u1) stable; urgency=medium

  [ jongkyung.woo ]
  * [93b3c04d] Released gooroom-3.0

 -- Gooroom Autobuilder <jenkins@gooroom.kr>  Wed, 06 Jul 2022 19:24:17 +0900

gooroom-os-protector (3.1.0+5.10.46~4+bpo10+1+grm2u1) stable; urgency=medium

  [ jongkyung.woo ]
  * [6c67a638] Released Version 2.5.3+5.9.6-1~bpo10+1+grm2u1 (2020-12-24)
  * [574a287c] Released Version 2.5.3+5.10.46-4~bpo10+1+grm2u1 (2021-08-18)

  [ Seunghun Han ]
  * [bed75727] Support Linux kernel 5.10.0
  * [0be37b12] Change version to 3.0.0.
  * [3a3a6905] Disable the NMI watchdog feature
  * [736285e3] Remove dependencies on System.map files
  * [5edc4281] Remove a workaround test feature
  * [c4d8c513] Remove compile warnings
  * [be5d64c3] Change version to 3.1.0

 -- Gooroom Autobuilder <jenkins@gooroom.kr>  Thu, 21 Oct 2021 18:29:02 +0900

gooroom-os-protector (2.5.5+4.19.194~3+grm2u1) stable; urgency=medium

  [ jongkyung.woo ]
  * [c6da21a6] Released Version 2.5.5+4.19.194~3+grm2u1 (2021-10-02)

 -- Gooroom Autobuilder <jenkins@gooroom.kr>  Sat, 02 Oct 2021 17:33:48 +0900

gooroom-os-protector (2.5.5+4.19.171~2+grm2u1) stable; urgency=medium

  [ Seunghun Han ]
  * [739b6703] Add a detecting Intel CPU feature
  * [ec7474cf] Change CR4 monitoring algorithm
  * [8c620e7b] Add hardening flags
  * [9563fad8] Change version to 2.5.5

 -- Gooroom Autobuilder <jenkins@gooroom.kr>  Sat, 02 Oct 2021 15:38:29 +0900

gooroom-os-protector (2.5.4+4.19.171~2+grm2u1) stable; urgency=medium

  [ jongkyung.woo ]
  * [b2d1d1de] Released Version 2.5.4+4.19.171-2+grm2u1 (2021-05-04)

 -- Gooroom Autobuilder <jenkins@gooroom.kr>  Thu, 06 May 2021 11:21:18 +0900

gooroom-os-protector (2.5.4+4.19.152~1+grm2u2) stable; urgency=medium

  [ Seunghun Han ]
  * [faf4ec8f] Add Celeron N4000 support feature and change to v2.5.4

  [ Gooroom Autobuilder ]

 -- Gooroom Autobuilder <jenkins@gooroom.kr>  Mon, 22 Feb 2021 12:12:35 +0900

gooroom-os-protector (2.5.3+4.19.152~1+grm2u2) stable; urgency=medium

  [ jongkyung.woo ]
  * [65bf97d7] Released Version 2.5.3+4.19.152~1+grm2u2 (2020-11-12)

  [ Gooroom Autobuilder ]

 -- Gooroom Autobuilder <jenkins@gooroom.kr>  Sat, 14 Nov 2020 13:13:08 +0900

gooroom-os-protector (2.5.3+4.19.152~1+grm2u1) stable; urgency=medium

  [ jongkyung.woo ]
  * [4403870a] Released Version 2.5.3+4.19.152~1+grm2u1 (2020-10-30)

  [ Gooroom Autobuilder ]

 -- Gooroom Autobuilder <jenkins@gooroom.kr>  Mon, 02 Nov 2020 16:38:06 +0900

gooroom-os-protector (2.5.3+4.19.118~2+deb10u1+grm2u1) stable; urgency=medium

  [ Seunghun Han ]
  * [51c994d7] Add reboot code for Samsung Notebook Issue
  * [9a001151] Change version to 2.5.3

  [ Gooroom Autobuilder ]

 -- Gooroom Autobuilder <jenkins@gooroom.kr>  Tue, 01 Sep 2020 14:29:03 +0900

gooroom-os-protector (2.5.2+4.19.118~2+deb10u1+grm2u3) stable; urgency=medium

  [ Seunghun Han ]
  * [de58f515] Add logind option for Samsung Notebook Issue

  [ jongkyung.woo ]
  * [1f13a346] Revert disable-hibernate.pkla and disable-suspend.pkla for Samsung Notebook Issue

  [ Gooroom Autobuilder ]

 -- Gooroom Autobuilder <jenkins@gooroom.kr>  Thu, 27 Aug 2020 22:39:36 +0900

gooroom-os-protector (2.5.2+4.19.118~2+deb10u1+grm2u2) stable; urgency=medium

  [ Seunghun Han ]
  * [804a3af7] Add reboot option for Samsung Notebook

  [ Gooroom Autobuilder ]

 -- Gooroom Autobuilder <jenkins@gooroom.kr>  Thu, 27 Aug 2020 17:13:20 +0900

gooroom-os-protector (2.5.2+4.19.118~2+deb10u1+grm2u1) stable; urgency=medium

  [ Seunghun Han ]
  * [f2ca4540] Fix an IDT entry size bug and enlarge a RO buffer size
  * [d8f6c92a] Remove useless boot option, intremap=off and prepare for 2.5.2 release

  [ Gooroom Autobuilder ]

 -- Gooroom Autobuilder <jenkins@gooroom.kr>  Tue, 25 Aug 2020 16:24:09 +0900

gooroom-os-protector (2.5.1+4.19.118~2+deb10u1+grm2u2) stable; urgency=medium

  [ Seunghun Han ]
  * [c8c9238a] Add kernel 4.9.0 support feature

  [ Gooroom Autobuilder ]

 -- Gooroom Autobuilder <jenkins@gooroom.kr>  Fri, 21 Aug 2020 17:59:00 +0900

gooroom-os-protector (2.5.1+4.19.118~2+deb10u1+grm2u1) stable; urgency=medium

  [ Seunghun Han ]
  * [7b6f9b1d] Add a new IDT for handling host's NMI interrupt and change version to 2.5.1

  [ Gooroom Autobuilder ]

 -- Gooroom Autobuilder <jenkins@gooroom.kr>  Wed, 05 Aug 2020 10:27:04 +0900

gooroom-os-protector (2.5.0+4.19.118~2+deb10u1+grm2u1) stable; urgency=medium

  [ Seunghun Han ]
  * [21d6aef7] Reduce the number and the range of locks
  * [4e348687] Manage pages instead of kmalloc and map all physical range
  * [89885db2] Fix uninteded context switching with mutex_unlock() and managing RO area bug
  * [fb22f657] Pass unintended interrupts including NMI to the normal world
  * [58f9f489] Add terminating malicious processes and modules feature
  * [3acd6504] Enhance a sleep feature for the unused case
  * [80f12794] Update version to 2.5.0 and change boot and Shadow-box options for it

  [ Gooroom Autobuilder ]

 -- Gooroom Autobuilder <jenkins@gooroom.kr>  Fri, 31 Jul 2020 11:40:15 +0900

gooroom-os-protector (2.4.1+4.19.118~2+deb10u1+grm2u1) stable; urgency=medium

  [ jongkyung.woo ]
  * [1fe5134c] Released Version 2.4.1+4.19.118_2+deb10u1+grm2u1

  [ Gooroom Autobuilder ]


Source/shadow_box/symbol.h
char* g_kernel_version[] = 
{
	"#1 SMP gooroom 4.7.2-1~bpo8+1+grm1 (2016-09-29)",
	"#1 SMP gooroom 4.7.5-1~bpo8+2+grm2 (2016-10-13)",
	"#1 SMP gooroom 4.5.4-1~bpo8+1+grm1 (2016-06-13)",
	"#1 SMP gooroom 3.16.7-ckt20-1+deb8u4+grm3 (2016-05-27)",
	"#1 SMP gooroom 4.9.30-2+deb9u2+grm1 (2017-07-04)",
	"#1 SMP gooroom 3.16.7-ckt20-1+deb8u4+grm2 (2016-05-10)",
};

struct sb_symbol_table_struct g_symbol_table_array[] =
{
	//#1 SMP gooroom 4.7.2-1~bpo8+1+grm1 (2016-09-29)
	{
		{
			{"_text", 0xffffffff81000000},
			{"walk_system_ram_range", 0xffffffff81083910},
			{"wake_up_new_task", 0xffffffff810a7300},
			{"free_module", 0xffffffff81100990},
			{"ftrace_module_init", 0xffffffff811317c0},
			{"proc_flush_task", 0xffffffff81269a00},
			{"secure_dccpv6_sequence_number", 0xffffffff814dfc60},
			{"secure_ipv4_port_ephemeral", 0xffffffff814dffb0},
			{"netif_rx_internal", 0xffffffff814e8bc0},
			{"__netif_receive_skb_core", 0xffffffff814e9050},
			{"netif_receive_skb_internal", 0xffffffff814e9b20},
			{"__ip_select_ident", 0xffffffff81523fe0},
			{"_etext", 0xffffffff815f61ab},
			{"__start___ex_table", 0xffffffff815f63b0},
			{"__stop___ex_table", 0xffffffff815f9578},
			{"__start_rodata", 0xffffffff81600000},
			{"__end_rodata", 0xffffffff818bf000},
			{"tasklist_lock", 0xffffffff81a05080},
			{"init_level4_pgt", 0xffffffff81a06000},
			{"modules", 0xffffffff81a5d970},
			{"init_mm", 0xffffffff81a74fc0},
			{"logbuf_lock", 0xffffffff81ccbdcc},
		},
	},
	//#1 SMP gooroom 4.7.5-1~bpo8+2+grm2 (2016-10-13)
	{
		{
			{"_text", 0xffffffff81000000},
			{"walk_system_ram_range", 0xffffffff81083910},
			{"wake_up_new_task", 0xffffffff810a7300},
			{"free_module", 0xffffffff81100990},
			{"ftrace_module_init", 0xffffffff811317c0},
			{"proc_flush_task", 0xffffffff81269a00},
			{"secure_dccpv6_sequence_number", 0xffffffff814dfc60},
			{"secure_ipv4_port_ephemeral", 0xffffffff814dffb0},
			{"netif_rx_internal", 0xffffffff814e8bc0},
			{"__netif_receive_skb_core", 0xffffffff814e9050},
			{"netif_receive_skb_internal", 0xffffffff814e9b20},
			{"__ip_select_ident", 0xffffffff81523fe0},
			{"_etext", 0xffffffff815f61ab},
			{"__start___ex_table", 0xffffffff815f63b0},
			{"__stop___ex_table", 0xffffffff815f9578},
			{"__start_rodata", 0xffffffff81600000},
			{"__end_rodata", 0xffffffff818bf000},
			{"tasklist_lock", 0xffffffff81a05080},
			{"init_level4_pgt", 0xffffffff81a06000},
			{"modules", 0xffffffff81a5d970},
			{"init_mm", 0xffffffff81a74fc0},
			{"logbuf_lock", 0xffffffff81ccbdcc},
		},
	},
	//#1 SMP gooroom 4.5.4-1~bpo8+1+grm1 (2016-06-13)
	{
		{
			{"_text", 0xffffffff81000000},
			{"walk_system_ram_range", 0xffffffff8107f9f0},
			{"wake_up_new_task", 0xffffffff810a3f30},
			{"free_module", 0xffffffff810f8420},
			{"ftrace_module_init", 0xffffffff81127210},
			{"proc_flush_task", 0xffffffff8125a680},
			{"secure_dccpv6_sequence_number", 0xffffffff814bdd60},
			{"secure_ipv4_port_ephemeral", 0xffffffff814be0b0},
			{"netif_rx_internal", 0xffffffff814c6790},
			{"__netif_receive_skb_core", 0xffffffff814c6c20},
			{"netif_receive_skb_internal", 0xffffffff814c76a0},
			{"__ip_select_ident", 0xffffffff81501110},
			{"_etext", 0xffffffff815d0bdf},
			{"__start___ex_table", 0xffffffff815d0df0},
			{"__stop___ex_table", 0xffffffff815d2df8},
			{"__start_rodata", 0xffffffff81600000},
			{"__end_rodata", 0xffffffff818e2000},
			{"tasklist_lock", 0xffffffff81a0b080},
			{"init_level4_pgt", 0xffffffff81a0c000},
			{"modules", 0xffffffff81a61370},
			{"init_mm", 0xffffffff81a773c0},
			{"logbuf_lock", 0xffffffff81ce4bb0},
		},
	},
	//#1 SMP gooroom 3.16.7-ckt20-1+deb8u4+grm3 (2016-05-27)
	{
		{
			{"_text", 0xffffffff81000000},
			{"walk_system_ram_range", 0xffffffff8106d870},
			{"wake_up_new_task", 0xffffffff81096c80},
			{"free_module", 0xffffffff810d8c60},
			{"ftrace_module_init", 0xffffffff811061a0},
			{"proc_flush_task", 0xffffffff8120bd80},
			{"secure_ipv4_port_ephemeral", 0xffffffff814200b0},
			{"secure_dccpv6_sequence_number", 0xffffffff814201e0},
			{"netif_rx_internal", 0xffffffff81426080},
			{"__netif_receive_skb_core", 0xffffffff81427090},
			{"netif_receive_skb_internal", 0xffffffff81427840},
			{"__ip_select_ident", 0xffffffff8145a5e0},
			{"_etext", 0xffffffff81529281},
			{"__start___ex_table", 0xffffffff81529460},
			{"__stop___ex_table", 0xffffffff8152b490},
			{"__start_rodata", 0xffffffff81600000},
			{"__end_rodata", 0xffffffff817cc000},
			{"tasklist_lock", 0xffffffff8180c040},
			{"init_level4_pgt", 0xffffffff81813000},
			{"modules", 0xffffffff81859090},
			{"init_mm", 0xffffffff81865700},
			{"logbuf_lock", 0xffffffff81a8d0d0},
		},
	},
	//#1 SMP gooroom 4.9.30-2+deb9u2+grm1 (2017-07-04)
	{
		{
			{"_text", 0xffffffff81000000},
			{"walk_system_ram_range", 0xffffffff8107e400},
			{"wake_up_new_task", 0xffffffff810a2af0},
			{"free_module", 0xffffffff810feae0},
			{"ftrace_module_init", 0xffffffff811313d0},
			{"proc_flush_task", 0xffffffff812775b0},
			{"secure_ipv4_port_ephemeral", 0xffffffff81511ed0},
			{"secure_dccpv6_sequence_number", 0xffffffff81512190},
			{"netif_rx_internal", 0xffffffff8151af50},
			{"__netif_receive_skb_core", 0xffffffff8151baa0},
			{"netif_receive_skb_internal", 0xffffffff8151c500},
			{"__ip_select_ident", 0xffffffff815592b0},
			{"_etext", 0xffffffff81631d82},
			{"__start___ex_table", 0xffffffff81631f90},
			{"__stop___ex_table", 0xffffffff8163508c},
			{"__start_rodata", 0xffffffff81800000},
			{"__end_rodata", 0xffffffff81ad2000},
			{"tasklist_lock", 0xffffffff81c050c0},
			{"init_level4_pgt", 0xffffffff81c07000},
			{"modules", 0xffffffff81c52b30},
			{"init_mm", 0xffffffff81c69d80},
			{"logbuf_lock", 0xffffffff81e9e308},
		},
	},
	//#1 SMP gooroom 3.16.7-ckt20-1+deb8u4+grm2 (2016-05-10)
	{
		{
			{"_text", 0xffffffff81000000},
			{"walk_system_ram_range", 0xffffffff8106c870},
			{"wake_up_new_task", 0xffffffff81095c80},
			{"free_module", 0xffffffff810d7c60},
			{"ftrace_module_init", 0xffffffff811051a0},
			{"proc_flush_task", 0xffffffff8120ad80},
			{"secure_ipv4_port_ephemeral", 0xffffffff8141f0b0},
			{"secure_dccpv6_sequence_number", 0xffffffff8141f1e0},
			{"netif_rx_internal", 0xffffffff81425080},
			{"__netif_receive_skb_core", 0xffffffff81426090},
			{"netif_receive_skb_internal", 0xffffffff81426840},
			{"__ip_select_ident", 0xffffffff814595e0},
			{"_etext", 0xffffffff81528281},
			{"__start___ex_table", 0xffffffff81528460},
			{"__stop___ex_table", 0xffffffff8152a490},
			{"__start_rodata", 0xffffffff81600000},
			{"__end_rodata", 0xffffffff817cc000},
			{"tasklist_lock", 0xffffffff8180c040},
			{"init_level4_pgt", 0xffffffff81813000},
			{"modules", 0xffffffff81858cd0},
			{"init_mm", 0xffffffff81865340},
			{"logbuf_lock", 0xffffffff81a8c0d0},
		},
	},
};

